src/main/java/org/broadinstitute/hellbender/utils/variant/writers/ShardingVCFWriter.java

```diff
@@ -0,0 +1,161 @@
+
+    public static final String SHARD_INDEX_PREFIX = ".shard_";
+    public static final String SHARD_INDEX_SUFFIX = FileExtensions.COMPRESSED_VCF;
+
+    private VariantContextWriter writer;
+    private VCFHeader header;
+    private final int maxVariantsPerShard;
+    private final Path basePath;
+    private final SAMSequenceDictionary dictionary;
+    private final boolean createMD5;
+    private final Options[] options;
+
+    /** Current shard  */
+    private int shardIndex;
+    /** Number of records written to current shard  */
+    private int shardSize;
+     /** Whether to write header, or null if header is undefined */
+    private Boolean enableWriteHeader;
+
+    /**
+     * Create a new sharding VCF writer
+     *
+     * @param basePath              base path of the output VCFs. The shard designation and file extension will be added.
+     * @param maxVariantsPerShard   max number of records per file (last shard may have less)
+     * @param dictionary            sequence dictionary for this writer
+     * @param createMD5             enable MD5 file creation
+     * @param options               vcf writer options
+     */
+    public ShardingVCFWriter(final Path basePath,
+                             final int maxVariantsPerShard,
+                             final SAMSequenceDictionary dictionary,
+                             final boolean createMD5,
+                             final Options... options) {
+        Utils.nonNull(basePath);
+        Utils.validateArg(maxVariantsPerShard > 0, "maxVariantsPerShard must be positive");
+        this.basePath = IOUtils.removeExtension(basePath, FileExtensions.VCF_LIST);
+        this.maxVariantsPerShard = maxVariantsPerShard;
+        this.dictionary = dictionary;
+        this.createMD5 = createMD5;
+        this.options = options;
+
+        // Initialize first shard
+        this.shardIndex = 0;
+        this.shardSize = 0;
+        this.writer = createNewWriter();
+    }
+
+    /**
+     * Initializes a new sharded file.
+     */
+    protected void createNextShard() {
+        writer.close();
+        shardIndex++;
+        shardSize = 0;
+        writer = createNewWriter();
+        Utils.nonNull(header, "Attempted to create new shard before header has been set");
+        initializeShardHeader();
+    }
+
+    /**
+     * Initializes shard header depending on which header function (set or write) was used
+     */
+    protected void initializeShardHeader() {
+        if (enableWriteHeader != null) {
+            if (enableWriteHeader.booleanValue()) {
+                writer.writeHeader(header);
+            } else {
+                writer.setHeader(header);
+            }
+        }
+    }
+
+    /**
+     * Creates a writer for a new shard
+     *
+     * @return the new writer
+     */
+    protected VariantContextWriter createNewWriter() {
+        final Path outPath = Paths.get(basePath + SHARD_INDEX_PREFIX + shardIndex + SHARD_INDEX_SUFFIX);
```

Comment:
I know we can't predict how many shards there will be, but it would be nice to zero pad this to four digits so the shards always sort nicely.

Reply:
Done (went to 5 zeros)